cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java

```diff
@@ -141,13 +170,20 @@ private void setResultSetMetadata() throws Exception {
                                                                dbSinkConfig.columns, dbSinkConfig.tableName))
       ) {
         ResultSetMetaData resultSetMetadata = rs.getMetaData();
-        int columnCount = resultSetMetadata.getColumnCount();
-        columnTypes = new int[columnCount];
         // JDBC driver column indices start with 1
-        for (int i = 0; i < columnCount; i++) {
-          columnTypes[i] = resultSetMetadata.getColumnType(i + 1);
+        for (int i = 0; i < rs.getMetaData().getColumnCount(); i++) {
+          String name = resultSetMetadata.getColumnName(i + 1);
+          int type = resultSetMetadata.getColumnType(i + 1);
+          columnToType.put(name, type);
         }
       }
+
+      columns = ImmutableList.copyOf(Splitter.on(",").split(dbSinkConfig.columns));
```

Comment:
can it move outside this try-catch too?

I mean after the connection is closed?